### PR TITLE
`SharedHostPinnedSpace` alias in fwd declaration

### PR DESCRIPTION
**Pull request by JBludau**
Will need a rebase after #5289 is merged -> done

At the moment there is no `SharedHostPinnedSpace` for `OpenMPTarget` and `OpenACC`.
The unittest checks the existance of the alias and asserts that the changes made in one ExecutionSpace are visible in the other.
Some of the examples/tutorials could be changed to use the new `SharedSpace` and `SharedHostPinnedSpace` -> separate pr

- [x] write documentation -> https://github.com/kokkos/kokkos-core-wiki/pull/177
- [ ] add to changelog

### CODE REVIEW

**Review by crtrott — CHANGES_REQUESTED** [review]
Needs the "does this exist" macro and constexpr bool.

File: core/src/Kokkos_Core_fwd.hpp [hunk 1]
```diff
@@ -196,7 +196,7 @@ using SharedSpace = Experimental::SYCLSharedUSMSpace;
 #define KOKKOS_HAS_SHARED_SPACE 1
 // if only host compile point to HostSpace
 #elif !defined(KOKKOS_ENABLE_OPENACC) && !defined(KOKKOS_ENABLE_OPENMPTARGET)
-using SharedSpace = HostSpace;
+using SharedSpace               = HostSpace;
```
Comment (JBludau) [inline core/src/Kokkos_Core_fwd.hpp:199]: not sure what clang-format is doing here ...

File: core/src/Kokkos_Core_fwd.hpp [hunk 2]
```diff
@@ -207,6 +207,27 @@ inline constexpr bool has_shared_space =
...
+#define KOKKOS_HAS_SHARED_HOST_PINNED_SPACE 1
+#elif defined(KOKKOS_ENABLE_HIP)
+using SharedHostPinnedSpace = HIPHostPinnedSpace;
+#define KOKKOS_HAS_SHARED_HOST_PINNED_SPACE 1
+#elif defined(KOKKOS_ENABLE_SYCL)
+    using SharedHostPinnedSpace = Experimental::SYCLHostUSMSpace;
+#define KOKKOS_HAS_SHARED_HOST_PINNED_SPACE 1
+#elif !defined(KOKKOS_ENABLE_OPENACC) && !defined(KOKKOS_ENABLE_OPENMPTARGET)
+    using SharedHostPinnedSpace = HostSpace;
+#define KOKKOS_HAS_SHARED_HOST_PINNED_SPACE 1
```
Comment (JBludau) [inline core/src/Kokkos_Core_fwd.hpp:221]: ... and here ...

**Comment by JBludau** [comment]
> Needs the "does this exist" macro and constexpr bool.

is added in [5ae6d4](https://github.com/kokkos/kokkos/pull/5405/commits/5ae6d4ff2fe2844c08e18dd2768a654c0bd46964)

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  IncrementFunctor() = delete;
+
+  explicit IncrementFunctor(ViewType view) : view_(view) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const index_type idx) const { ++view_(idx); }
+};
+
+template <typename ExecSpace, typename ViewType>
+void incrementInLoop(ViewType& view) {
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:65]: Consider moving this into the constructor of `IncrementFunctor` (would give it another name if you go that direction)

Comment (dalg24) [reply core/unit_test/TestSharedHostPinnedSpace.cpp:65]: Also pass in an execution space and so give it to the range policy to enable template argument deduction.

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+*/
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+
+namespace {
+
+template <typename ViewType>
+class IncrementFunctor {
+ private:
+  using index_type = decltype(std::declval<ViewType>().size());
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:52]: Complicated way to deduce `size_t`

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+  explicit IncrementFunctor(ViewType view) : view_(view) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const index_type idx) const { ++view_(idx); }
+};
+
+template <typename ExecSpace, typename ViewType>
+void incrementInLoop(ViewType& view) {
+  using index_type = decltype(view.size());
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:66]: Assuming you would specify anything else than View::size_type aka size_t, it is not propagated to the functor and the index are casted to the `IncrementFunctor::index_type`.  My point is yo0u write more but do not actually enable customization, and even you were doing it properly I am not convinced it is useful.

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+ public:
+  CheckFunctor() = delete;
+
+  CheckFunctor(ViewType view, int targetVal)
+      : view_(view), targetVal_(targetVal) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const index_type idx, bool& equality) const {
+    (view_(idx) == targetVal_) ? equality &= true : equality &= false;
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:93]: Yuk yuk yuk
Prefer `bool& equality -> int& err`
and
```C++
if (view_(idx) != targetVal_) { ++err; }

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 3 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: done in [ab0f14](https://github.com/kokkos/kokkos/pull/5405/commits/ab0f14be7e2cc2777de17625a9b32f9b40e7f3b1)

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 4 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: removed in [ab0f14](https://github.com/kokkos/kokkos/pull/5405/commits/ab0f14be7e2cc2777de17625a9b32f9b40e7f3b1)

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 5 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: right ... removed in [ab0f14](https://github.com/kokkos/kokkos/pull/5405/commits/ab0f14be7e2cc2777de17625a9b32f9b40e7f3b1)

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 6 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: changed in [ab0f14](https://github.com/kokkos/kokkos/pull/5405/commits/ab0f14be7e2cc2777de17625a9b32f9b40e7f3b1)

File: core/unit_test/CMakeLists.txt [hunk 7]
```diff
@@ -823,7 +824,10 @@ SET(DEFAULT_DEVICE_SOURCES
 # FIXME_OPENMPTARGET and FIXME_OPENACC do not provide a MemorySpace that can be accessed from all ExecSpaces
 # FIXME_SYCL clock_tic does not give the correct timings for cloc_tic
 if ((KOKKOS_ENABLE_OPENACC) OR (KOKKOS_ENABLE_OPENMPTARGET) OR (KOKKOS_ENABLE_SYCL))
```
Comment (dalg24) [inline core/unit_test/CMakeLists.txt:826]:
```suggestion
if (KOKKOS_ENABLE_OPENACC OR KOKKOS_ENABLE_OPENMPTARGET OR KOKKOS_ENABLE_SYCL)
```

Comment (dalg24) [reply core/unit_test/CMakeLists.txt:826]: I suppose this is not part of your patch so you can ignore

**Review by crtrott — APPROVED** [review]

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+//
+// ************************************************************************
+//@HEADER
+*/
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+
+namespace {
+
+template <typename ExecutionSpace, typename ViewType>
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:49]: No, template the constructor, not the class

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  if (!passed) {
+    std::cout << "Allocating " << numInts
+              << " ints in SharedHostPinnedSpace.\n";
+
+    std::cout << "Behavior found: \n";
+    std::cout
+        << "SharedHostPinnedSpace is read/write accessible by host and device: "
+        << readWriteAccessible << ", we expect true \n\n";
+  }
+  ASSERT_TRUE(passed);
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:139]:
```suggestion
  ASSERT_TRUE(readWriteAccessible) << "PRINT ON FAILURE";
```

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 8 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: done in [f7ee93](https://github.com/kokkos/kokkos/pull/5405/commits/f7ee93adc81f0df5297721ee7ce8a522a002e85f)

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 9 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: done in [d97863](https://github.com/kokkos/kokkos/pull/5405/commits/d9786320d3f5711f8b339c39bcf0e854fa5b64c1)

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  KOKKOS_FUNCTION
+  void operator()(const size_t idx) const { ++view_(idx); }
+};
+
+template <typename ViewType>
+struct CheckResult {
+  ViewType view_;
+  const int targetVal_;
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:71]:
```suggestion
  int targetVal_;
```

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+  unsigned error = 0;
+
+  template <typename ExecutionSpace>
+  CheckResult(ExecutionSpace, ViewType view, int targetVal)
+      : view_(view), targetVal_(targetVal) {
+    Kokkos::parallel_reduce(
+        "check",
+        Kokkos::RangePolicy<ExecutionSpace, Kokkos::IndexType<size_t>>{
+            0, view_.size()},
+        *this, Kokkos::Sum<unsigned>(error));
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:81]: Does CTAD not kick in?
```suggestion
        *this, Kokkos::Sum(error));
```

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+  KOKKOS_FUNCTION
+  void operator()(const size_t idx) const { ++view_(idx); }
+};
+
+template <typename ViewType>
+struct CheckResult {
+  ViewType view_;
+  const int targetVal_;
+  unsigned error = 0;
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:72]: I would slightly prefer saying `errors` plural

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    Kokkos::parallel_reduce(
+        "check",
+        Kokkos::RangePolicy<ExecutionSpace, Kokkos::IndexType<size_t>>{
+            0, view_.size()},
+        *this, Kokkos::Sum<unsigned>(error));
+    Kokkos::fence();
+  }
+
+  KOKKOS_FUNCTION
+  void operator()(const size_t idx, unsigned& error_val) const {
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:86]: `error_val` is the only occurrence in the file of your mixing snake_case.  I would prefer `localError` or `err`

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+  // ALLOCATION
+  Kokkos::View<int*, Kokkos::SharedHostPinnedSpace> sharedData("sharedData",
+                                                               numInts);
+  // MAIN LOOP
+  unsigned incrementCount = 0;
+
+  for (unsigned i = 0; i < numDeviceHostCycles; ++i) {
+    // INCREMENT DEVICE
+    Increment incrementDevice(Kokkos::DefaultExecutionSpace{}, sharedData);
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:110]: I would prefer if you defined `deviceExecSpace` and `hostExecSpace` and did
```C++
Increment(deviceExecSpace, sharedData);
...
CheckResult(hostExecSpace, sharedData, incrementCount);
...
Increment(hostExecSpace, sharedData);
...
CheckResult(deviceExecSpace, sharedData, incrementCount);
```

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    ++incrementCount;
+    // CHECK RESULTS HOST
+    CheckResult checkHost(Kokkos::DefaultHostExecutionSpace{}, sharedData,
+                          incrementCount);
+    ASSERT_TRUE(checkHost.error == 0)
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:115]:
```suggestion
    ASSERT_EQ(checkHost.error, 0)
```
Same comment below.  That is the only one that you must fix.  Others I will leave up to you.

File: core/src/Kokkos_Core_fwd.hpp [hunk 16]
```diff
@@ -196,7 +196,7 @@ using SharedSpace = Experimental::SYCLSharedUSMSpace;
 #define KOKKOS_HAS_SHARED_SPACE 1
```
Comment (masterleinad) [inline core/src/Kokkos_Core_fwd.hpp:196]: Would you remind me why we gave this a value instead of using a simple
```
#define KOKKOS_HAS_SHARED_SPACE
```
?

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 11 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]:
```
/kokkos/core/unit_test/TestSharedHostPinnedSpace.cpp:81:16: error: no viable constructor or deduction guide for deduction of template arguments of 'Sum'
        *this, Kokkos::Sum(numErrors));
```

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 12 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: done in [593cb4](https://github.com/kokkos/kokkos/pull/5405/commits/593cb4571007bd8c1bbcc6b85dda4d23017bf1e4)

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 13 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: done in [593cb4](https://github.com/kokkos/kokkos/pull/5405/commits/593cb4571007bd8c1bbcc6b85dda4d23017bf1e4)

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 14 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: done in [0fac2a](https://github.com/kokkos/kokkos/pull/5405/commits/0fac2a538480266303a158ffec5cb8c52f04962b)

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 15 above)
Comment (JBludau) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: done in [0fac2a](https://github.com/kokkos/kokkos/pull/5405/commits/0fac2a538480266303a158ffec5cb8c52f04962b)

File: core/src/Kokkos_Core_fwd.hpp (hunk 16 above)
Comment (JBludau) [reply core/src/Kokkos_Core_fwd.hpp:196]: my reasoning was to allow using it inside an `if` ... even though we have the `constexp` var

https://godbolt.org/z/TdoWfrMoG

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 11 above)
Comment (masterleinad) [reply core/unit_test/TestSharedHostPinnedSpace.cpp]: All built-in reducers have also a memory space template argument.

File: core/unit_test/TestSharedHostPinnedSpace.cpp [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+
+  template <typename ExecutionSpace>
+  CheckResult(ExecutionSpace, ViewType view, int targetVal)
+      : view_(view), targetVal_(targetVal) {
+    Kokkos::parallel_reduce(
+        "check",
+        Kokkos::RangePolicy<ExecutionSpace, Kokkos::IndexType<size_t>>{
+            0, view_.size()},
+        *this, Kokkos::Sum<unsigned>(numErrors));
+    Kokkos::fence();
```
Comment (dalg24) [inline core/unit_test/TestSharedHostPinnedSpace.cpp:82]: Why did you add a fence by the way?

**Review by dalg24 — APPROVED** [review]

File: core/src/Kokkos_Core_fwd.hpp (hunk 16 above)
Comment (dalg24) [reply core/src/Kokkos_Core_fwd.hpp:196]: I don't like it but this was already there.  Maybe we can discuss at the developer meeting today and possibly update in a follow up PR.

Comment (masterleinad) [reply core/src/Kokkos_Core_fwd.hpp:196]: Sounds good.

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 17 above)
Comment (crtrott) [reply core/unit_test/TestSharedHostPinnedSpace.cpp:82]: That fence is not needed.

**Comment by crtrott** [comment]
Retest this please.

File: core/unit_test/TestSharedHostPinnedSpace.cpp (hunk 17 above)
Comment (dalg24) [reply core/unit_test/TestSharedHostPinnedSpace.cpp:82]: None of the two fences are

**Review by masterleinad — APPROVED** [review]

**Comment by masterleinad** [comment]
CUDA-10.1-Clang-Tidy shows:
```
[ RUN      ] defaultdevicetype.shared_host_pinned_space
11: /var/jenkins/workspace/Kokkos/core/unit_test/TestSharedHostPinnedSpace.cpp:115: Failure
11: Expected equality of these values:
11:   CheckResult(HostExecutionSpace{}, sharedData, incrementCount).numErrors
11:     Which is: 1024
11:   0u
11:     Which is: 0
```

**Comment by JBludau** [comment]
Retest this please

**Comment by JBludau** [comment]
Retest this please
